core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -630,13 +630,17 @@ class ReplicaManager(val config: KafkaConfig,
 
     // Third delete the logs and checkpoint.
     val errorMap = new mutable.HashMap[TopicPartition, Throwable]()
+    // `tieredEnabledPartitions` has exclude internal topics
+    val tieredEnabledPartitions = partitionsToStop.filter(sp => logManager.getLog(sp.topicPartition).exists(unifiedLog => unifiedLog.remoteLogEnabled()))
```

Comment:
This could be moved after the if {} logic in next line and merged after `partitionsToStop.filterNot(sp => errorMap.contains(sp.topicPartition))`. This way you won't call a contains() operation per partition. 

Also:
1. we want to exclude reserved/internal topics as well.
2. we want to add back the warn that we talked about in https://github.com/apache/kafka/pull/14329#discussion_r1315675510 since we no longer expect stopPartitions to be called for non-tiered and reserved topic.

Reply:
The calculation of `tieredEnabledPartitions` may not be placed after if{}, because `logManager.asyncDelete` will remove the corresponding unifiedLog from `currentLogs` in `logManager`, causing the calculation of `tieredEnabledPartitions` to be incorrect.
1. The calculation of `tieredEnabledPartitions` actually has excluded internal topics, because (here)[https://github.com/apache/kafka/blob/8f8ad6db384ce30e8a5d848d3ed826a3f7a54dfe/core/src/main/scala/kafka/log/UnifiedLog.scala#L1887]
2. added.
@divijvaidya